lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -119,7 +129,9 @@ static bool hasInfinitelyRecursiveApply(SILFunction *targetFn) {
 
     // If this block doesn't have a recursive call, and it exits the function,
     // then we know the function is not infinitely recursive.
-    if (curBlock->getTerminator()->isFunctionExiting())
+    TermInst *terminator = curBlock->getTerminator();
+    if (terminator->isFunctionExiting() ||
+        terminator->getTermKind() == TermKind::UnreachableInst)
```

Comment:
If you are going to go this way can you move this into a separate query function and exhaustive switch over term kind so that if people add a new weird terminator they know to update this?